source/common/common/callback_impl.h

```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
   const std::shared_ptr<bool> still_alive_{std::make_shared<bool>(true)};
 };
 
+/**
+ * Utility class for managing callbacks. Thread-safe.
+ * @see CallbackManager
+ */
+template <typename... CallbackArgs> class ThreadSafeCallbackManager {
```

Comment:
🤔  Hmm... the drain manager is the central object that is currently shared across threads, so my intention was to create a version of the `CallbackManager` that was thread-safe so that it could be used within the shared drain-manager object.

Open to another approach, but currently not sure how to go about it.

Reply:
So for the moment I've worked around this limitation by wrapping the callback body in a dispaterhcer `post()` function which is thread-safe.

Reply:
Resolving this thread post-refactor as they are no longer relevant.